### PR TITLE
Aditya 30 recipes

### PR DESCRIPTION
**Pull request by adityaguptakd**

### CODE REVIEW

**Comment by nandantumu** [comment]
Fails automated checks.

File: ceaos/actuators/actuator_commands.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from ..objects import Farm, Environment, Bed
```
Comment (nandantumu) [inline ceaos/actuators/actuator_commands.py:1]: Is this file just to provide stubs for the rest of the code? Like fake actuators to use? Why not use artificial actuators here instead?

File: ceaos/cfg.py [hunk 2]
```diff
@@ -204,6 +204,10 @@ def load_config(config_folder="ceaos.resources", config_file="config.yaml"):
                 bed_object = Bed(str(bed.get("name")))
                 add_sensors(bed_object, bed, sensors)
 
+                if bed.get("name") == "bed1":
+                    bed_object.add_actuators("158.130.113.133", "pH")
+                    bed_object.add_actuators("158.130.113.133", "EC")
```
Comment (nandantumu) [inline ceaos/cfg.py:209]: Why hardcoding the actuators?

File: ceaos/grow.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    for bed in bed_list:
+        for act in bed.get_actuators():
+            if actuator in bed.get_actuators()[act]:
+                ip_list.append(act)
+
+    return ip_list
+
+
+def send_command(ip_list, max, min, actuation):
+    for ip in ip_list:
```
Comment (nandantumu) [inline ceaos/grow.py:43]: This loop will send a command to every IP in the list. Sounds a bit much, right?

File: ceaos/resources/config/config_lettuce_grow.yml [hunk 4]
```diff
@@ -0,0 +1,87 @@
+plant: "lettuce"
```
Comment (nandantumu) [inline ceaos/resources/config/config_lettuce_grow.yml:1]: This and the template look good. Are they the same as Sarah's updated version?

File: scripts/runner.py [hunk 5]
```diff
@@ -51,6 +52,8 @@ def log_data(refresh_rate, logger_list, client):
     logging.info("DB Client Configured")
     farm_name = farm.get_name()
 
+    recipe_list = load_grow(farm, "ceaos.resources.config", "config_lettuce_grow.yml")
```
Comment (nandantumu) [inline scripts/runner.py:55]: This is a perfectly valid paradigm for loading grow recipes, but what happens when we have multiple, or different grow recipies for different beds or environments?

File: ceaos/objects/beds.py [hunk 6]
```diff
@@ -24,6 +24,18 @@ def delete_plant(self, plant):
         del self.plants[plant.name]
         self.num_plants -= 1
 
+    def recv_details(self, **kwargs):
```
Comment (nandantumu) [inline ceaos/objects/beds.py:27]: I think I'm unclear as to the intent of this function. Seems like the data must already be received for this function to work?

File: ceaos/objects/beds.py [hunk 7]
```diff
@@ -32,8 +44,12 @@ def delete_sensor(self, name):
         del self.sensors[name]
         self.num_sensors -= 1
 
-    def add_actuators(self, name, actuator):
-        self.actuators[name] = actuator
+    def add_actuators(self, ip, actuator):
```
Comment (nandantumu) [inline ceaos/objects/beds.py:47]: Why is IP the key for actuators, and not the name?

**Review by nandantumu — CHANGES_REQUESTED** [review]
Hey, I think I'm pretty unclear on a lot of this, and could do with a walkthrough. I'll reach out to set up some time to talk about it.

File: scripts/runner.py (hunk 5 above)
Comment (adityaguptakd) [reply scripts/runner.py:55]: not sure how to go about this

File: ceaos/resources/config/config_lettuce_grow.yml [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+environments:
+  - name: "env1"
+beds:
+  - name: "bed1"
```
Comment (nandantumu) [inline ceaos/resources/config/config_lettuce_grow.yml:5]: Let's try to move these to a new file to keep everything compartmentalized.

File: ceaos/resources/config.yaml [hunk 9]
```diff
@@ -10,6 +10,11 @@ environments: #environments are a list of dictionaries
         - type: "reservoir"
         - type: "camera"
         - type: "light"
+      actuators:
+        - type: "Artificial EC"
+          ip: 158.130.113.109
```
Comment (nandantumu) [inline ceaos/resources/config.yaml:15]: Let's remove the IPs for now

File: ceaos/resources/config/template.yml [hunk 10]
```diff
@@ -0,0 +1,86 @@
+plant: "type of plant being grow"
+environments:
+  - name: "Environments of interest"
+beds:
+  - name: "beds within each environment"
```
Comment (nandantumu) [inline ceaos/resources/config/template.yml:5]: Move as per other comment

File: scripts/runner.py (hunk 5 above)
Comment (nandantumu) [reply scripts/runner.py:55]: You should move the recipes to a subfolder in config, and just generate a list of the recipes. In the config file, you can parse the references to the recipes, and apply them after/during the config loading process.

File: ceaos/actuators/sample_actuator.py [hunk 11]
```diff
@@ -10,11 +10,14 @@ def __init__(self):
         self.running = None
         self.curr_state = None
         self.datatype = None
-    def set_point(self, sensor):
+        self.location = None
+        self.ip = None
```
Comment (nandantumu) [inline ceaos/actuators/sample_actuator.py:14]: Let's drop this for now

File: ceaos/actuators/sample_actuator.py [hunk 12]
```diff
@@ -10,11 +10,14 @@ def __init__(self):
...
+        self.sensor = None
+
+    def set_point(self, min, max):
```
Comment (nandantumu) [inline ceaos/actuators/sample_actuator.py:17]: Let's have more discussion on this method

File: ceaos/actuators/sample_actuator.py [hunk 13]
```diff
@@ -40,3 +43,21 @@ def get_datatype(self):
...
+        self.ip = ip
+
+    def get_ip(self, ip):
+        return self.ip
+
+    def set_sensor(self, sensor):
+        self.sensor = sensor
+
+    def get_sensor(self, sensor):
+        return self.sensor
```
Comment (nandantumu) [inline ceaos/actuators/sample_actuator.py:63]: Drop these methods

File: ceaos/actuators/sample_actuator.py [hunk 14]
```diff
@@ -10,11 +10,14 @@ def __init__(self):
         self.running = None
         self.curr_state = None
         self.datatype = None
-    def set_point(self, sensor):
+        self.location = None
+        self.ip = None
+        self.sensor = None
```
Comment (nandantumu) [inline ceaos/actuators/sample_actuator.py:15]: Let's drop this sensor

File: ceaos/cfg.py [hunk 15]
```diff
@@ -157,6 +158,23 @@ def error_handle(
...
+                a = Artificial_Actuator()
+            else:
+                print("network actuator not supported yet")
+
+            a.set_datatype(
+                actuator.get("type"))  # Sets type of data sensor is collecting
+            a.set_location(str(dictionary.get(
+                "name")))  # sets location to the name of the environment
+            if "ip" in actuator:
+                a.set_ip(actuator.get("ip"))
```
Comment (nandantumu) [inline ceaos/cfg.py:174]: drop

File: ceaos/cfg.py [hunk 16]
```diff
@@ -191,6 +209,7 @@ def load_config(config_folder="ceaos.resources", config_file="config.yaml"):
         return None, None, None, error
     else:
         sensors = []  # List of sensors
+        actuators = [] # List of actuators
```
Comment (nandantumu) [inline ceaos/cfg.py:212]: drop

File: ceaos/cfg.py [hunk 17]
```diff
@@ -218,7 +237,7 @@ def load_config(config_folder="ceaos.resources", config_file="config.yaml"):
 
         print("FARM SETUP COMPLETE")
 
-        return farm_object, sensors, connection_dict, error
+        return farm_object, sensors, actuators, connection_dict, error
```
Comment (nandantumu) [inline ceaos/cfg.py:240]: drop

File: ceaos/grow.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+from datetime import datetime, time
+from .objects.farm import Farm
+from .objects.beds import Bed
+from .objects.environment import Environment
+from .actuators.sample_actuator import Artificial_Actuator
+from importlib_resources import files
+
+
+def get_relevant_beds(farm, env_list, bed_names_list):
+    all_beds = dict()
```
Comment (nandantumu) [inline ceaos/grow.py:11]: rename variable to relevant_beds or something

**Review by nandantumu — CHANGES_REQUESTED** [review]

File: ceaos/objects/beds.py [hunk 19]
```diff
@@ -32,8 +32,8 @@ def delete_sensor(self, name):
         del self.sensors[name]
         self.num_sensors -= 1
 
-    def add_actuators(self, name, actuator):
-        self.actuators[name] = actuator
+    def add_actuators(self, name, new_actuator):
+        self.actuators[name] = new_actuator
         self.num_actuators += 1
```
Comment (nandantumu) [inline ceaos/objects/beds.py:37]: I would drop this num pattern, in python, you can always call `len(actuators)` and get the number of actuators inside.

**Review by nandantumu — CHANGES_REQUESTED** [review]
Almost there, one thing to change

**Review by nandantumu — APPROVED** [review]

**Pull request merged**